lib/src/canonicalized_map.dart

```diff
@@ -30,8 +30,8 @@ class CanonicalizedMap<C, K, V> implements Map<K, V> {
   /// The [isValidKey] function is called before calling [canonicalize] for
   /// methods that take arbitrary objects. It can be used to filter out keys
   /// that can't be canonicalized.
-  CanonicalizedMap(C Function(K key) canonicalize,
-      {bool Function(Object key) isValidKey})
+  CanonicalizedMap(C Function(K? key) canonicalize,
```

Comment:
This is unfortunate since it makes everyone using it have to handle nulls even if they do't want to. The `K` could already be nullable, why do we need it to be nullable here?

Reply:
Because of this comment on the class:

```
/// By default, `null` is allowed as a key. It can be forbidden via the
/// `isValidKey` parameter.
```

We could change that behavior though potentially?

Reply:
We should consider it. The UX is likely a lot worse this way. I would guess a lot of usages in the wild don't bother handling null keys because they just never come up.

Reply:
It was pretty explicitly handled here too https://github.com/dart-lang/collection/blob/master/lib/src/canonicalized_map.dart#L182.

But, we could just remove the explicit null check there and only leave the `is K` check.

This would be a more breaking change I think though.

Reply:
I think it was handled there because `null is K` behaved weirdly before.

With null safety you can always make `K` nullable if you want that behavior.

Reply:
`null is T` would have always been false before (unless T was Null) afaik? It does seem like it was an intentional choice to me but ya I don't really understand why.

Reply:
`null as T` would have worked though. I think it was an because `null` was always allowed as an argument for any `T`.